databricks/koalas/frame.py

```diff
@@ -5631,6 +5631,82 @@ def head(self, n: int = 5) -> "DataFrame":
                 sdf = sdf.orderBy(NATURAL_ORDER_COLUMN_NAME)
             return DataFrame(self._internal.with_new_sdf(sdf.limit(n)))
 
+    def last(self, offset) -> "DataFrame":
+        """
+        Select final periods of time series data based on a date offset.
+
+        When having a DataFrame with dates as index, this function can
+        select the last few rows based on a date offset.
+
+        Parameters
+        ----------
+        offset : str or DateOffset
+            The offset length of the data that will be selected. For instance,
+            '3D' will display all the rows having their index within the last 3 days.
+
+        Returns
+        -------
+        DataFrame
+            A subset of the caller.
+
+        Raises
+        ------
+        TypeError
+            If the index is not  a :class:`DatetimeIndex`
+
+        Examples
+        --------
+
+        >>> index = pd.date_range('2018-04-09', periods=4, freq='2D')
+        >>> pdf = pd.DataFrame({'A': [1, 2, 3, 4]}, index=index)
+        >>> kdf = ks.from_pandas(pdf)
+        >>> kdf
+                    A
+        2018-04-09  1
+        2018-04-11  2
+        2018-04-13  3
+        2018-04-15  4
+
+        Get the rows for the last 3 days:
+
+        >>> kdf.last('3D')
+                    A
+        2018-04-13  3
+        2018-04-15  4
+
+        Notice the data for 3 last calendar days were returned, not the last
+        3 observed days in the dataset, and therefore data for 2018-04-11 was
+        not returned.
+        """
+        # Check index type should be format DateTime
+        from databricks.koalas.indexes import DatetimeIndex
+
+        if not isinstance(self.index, DatetimeIndex):
+            raise TypeError("'last' only supports a DatetimeIndex")
+
+        offset = pd.tseries.frequencies.to_offset(offset)
+        index_max = self.index.max()
+        from_date = index_max - offset
+
+        kdf = self.copy()
+        kdf.index.name = verify_temp_column_name(kdf, "__index_name__")
+
+        def pandas_loc(pdf):
+            return pdf.loc[from_date:index_max].reset_index()
+
+        # apply_batch will remove the index of the Koalas DataFrame and attach a default index,
+        # which will never be used. So use "distributed" index as a dummy to avoid overhead.
+        with option_context("compute.default_index_type", "distributed"):
+            kdf = kdf.koalas.apply_batch(pandas_loc)
+
+        return DataFrame(
+            self._internal.copy(
+                spark_frame=kdf._internal.spark_frame,
+                index_spark_columns=kdf._internal.data_spark_columns[:1],
+                data_spark_columns=kdf._internal.data_spark_columns[1:],
+            )
+        )
```

Comment:
Did you intend not to use `return self[from_date:index_max]`?

Reply:
My original implementation did something similar `return self.loc[from_date:index_max]`, however, build failed due to `Incompatible return value type (got "Union[Series[Any], DataFrame[Any]]", expected "DataFrame[Any]")`

I changed implementation to ensure return type matched the expected. I definitely prefer the original implementation, however, struggled finding a way to meet the build requirement

Reply:
Apologies I've just pushed a change that follows similar implementation to the original and what you suggested and avoids `mypy` build errors.

As we only require the `from_date` we can filter from the `[from_date:]` rather than filtering between a range `[from_date:index_max]`

Note: `self[from_date:index_max]` throws `mypy` error `Slice index must be an integer or None` our index is a `Timestamp` and would want to avoid type conversions.

Reply:
Thanks for sharing your investigation!

It is quite helpful to know about the mypy slice check. Your current approach is smart. :)